testing.md

```diff
@@ -34,6 +34,82 @@ A piece of code is not done until it's tested.
 
 So just write tests. Always.
 
+## Testing Principles
+
+Things you should keep in mind while writing tests for a codebase.
+
+### Write Complete Testing Suites
+
+Testing suites should attempt to be exhaustive in verifying the code does what is expected. None of
+these tests should require any manual work, other than executing the initial command to run the
+tests.
+
+An exhaustive test suite implies that you should be confident of merging and deploying your changes
+once all your tests pass in your continuous integration environment. Any manual steps that are needed
+to verify code changes should be encoded as automated tests that run in your continuous integration
+environment.  As an example, if to test your code, you bring up an instance of your service and
```

Comment:
continuous integration is introduced at the end of the file - feels a bit confusing to bring it up before its defined